pkg/auth/multiauthenticator.go

```diff
@@ -5,64 +5,122 @@ import (
+// Use this implementation when the audience is expected to contain only the client_id.
+//
+// 2> iteratingMultiAuthenticator allows the audience value to be an array of strings. It delegates the interpretation
+// of audience to the specified athenticators by iterating over them. All authenticators use the same issuer and
+// at most one of the authenticators shall successfully authenticate the token. Others will reject the token
+// because the client_id will not match.
+
+type multiAuthenticatorByClientID struct {
+	issuer         string
+	authenticators map[string]Authenticator
 }
 
-// NewMultiAuthenticator maintains a list of authenticators for a given issuer.
-// The input argument is a map of issuers to a list of authenticators for that issuer.
-// NOTE: This interface does not check if there are duplicate authenticators. The interface
-// will own the input map and will not create a copy of it.
-func NewMultiAuthenticator(
-	authenticators map[string][]Authenticator,
-) (MultiAuthenticatorWithClientID, error) {
-	for issuer, authenticatorsList := range authenticators {
-		if len(authenticatorsList) == 0 {
-			return nil, fmt.Errorf("empty authenticators list for issuer %v", issuer)
-		}
+// NewMultiAuthenticatorByClientID returns Authenticator implementation that assumes that the audience
+// field in the token contains just the client ID, which is also the key in the authenticators map
+// passed to this function. All authenticators must use the same issuer.
+func NewMultiAuthenticatorByClientID(
+	issuer string,
+	authenticatorsByClientID map[string]Authenticator,
+) (Authenticator, error) {
+	if len(authenticatorsByClientID) == 0 {
+		return nil, fmt.Errorf("empty authenticators for issuer %s", issuer)
 	}
-	return &multiAuthenticatorImpl{
-		authenticators: authenticators,
+	return &multiAuthenticatorByClientID{
+		issuer:         issuer,
+		authenticators: authenticatorsByClientID,
 	}, nil
 }
 
-func (m *multiAuthenticatorImpl) GetAuthenticators(issuer string) []Authenticator {
-	return m.authenticators[issuer]
+func (m *multiAuthenticatorByClientID) AuthenticateToken(ctx context.Context, idToken string) (*Claims, error) {
+	tokenClaims, err := TokenClaims(idToken)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get claims from token: %w", err)
+	}
+	if tokenClaims.Issuer != m.issuer {
+		return nil, fmt.Errorf("token was issued by %s which does not match issuer %s of this multi authenticator",
+			tokenClaims.Issuer, m.issuer)
+	}
+	var clientID string
+	aud, err := tokenClaims.GetAudience()
+	if err != nil {
+		return nil, fmt.Errorf("failed to find audience in the claims: %w", err)
+	}
+	// This implementation assumes that the audience is a single string containing the client ID.
+	// If that is not true, the "iterating" multi authenticator should be used instead.
+	if len(aud) != 1 {
+		return nil, fmt.Errorf("unable to determine client ID because token has multiple audiences: %v", aud)
```

Comment:
May want to add in the comments in line 38 that this implementation is for tokens with only 1 value in the audience.

Reply:
Also, may want to have this set to `len(aud) > 1` instead to match the error string, and maybe add a `len(aud) == 0` for another error